datanode/snapshot/service_create_snapshot.go

```diff
@@ -0,0 +1,257 @@
+		if err != nil {
+			b.logError(fmt.Errorf("failed to compress history snapshot:%w", err))
+			return
+		}
+		b.log.Infof("compressed history snapshot data size: %dB", compressedHistoryByteCount)
+
+		if err := os.RemoveAll(currentDataDir); err != nil {
+			b.logError(fmt.Errorf("failed to remove uncompressed current data directory  %s :%w", currentDataDir, err))
+			return
+		}
+
+		if err := os.RemoveAll(historyDataDir); err != nil {
+			b.logError(fmt.Errorf("failed to remove uncompressed current data directory  %s :%w", historyDataDir, err))
+			return
+		}
+
+		err = os.Remove(writeLockFileName)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to remove write lock file:%w", err))
+			return
+		}
+
+		snapshotHash, err := GetSnapshotMd5Hash(b.log, meta)
+		if err != nil {
+			b.logError(fmt.Errorf("failed to get md5 hash of snapshot:%w", err))
+		}
+
+		metrics.SetLastSnapshotRowcount(float64(rowsCopied))
+		metrics.SetLastSnapshotCurrentStateBytes(float64(compressedCurrentStateByteCount))
+		metrics.SetLastSnapshotHistoryBytes(float64(compressedHistoryByteCount))
+		metrics.SetLastSnapshotSeconds(time.Since(start).Seconds())
+
+		b.log.Infof("finished creating snapshot for chain:%s, height:%d, time taken:%s, rows copied:%d, total compressed data size:%d, md5 hash:%s", chainID, toHeight,
+			time.Since(start), rowsCopied, compressedCurrentStateByteCount+compressedHistoryByteCount, snapshotHash)
+	}()
+
+	return meta, nil
+}
+
+func (b *Service) logError(err error) {
+	if b.config.PanicOnError {
+		panic(err)
+	}
+
+	b.log.Errorf(":%w", err)
+}
+
+func GetHistoryMd5Hash(log *logging.Logger, snapshot Meta) (string, error) {
+	snapshotHash := md5.New()
+
+	err := hashFile(log, snapshotHash, snapshot.HistorySnapshotFile)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(snapshotHash.Sum(nil)), nil
+}
+
+func GetSnapshotMd5Hash(log *logging.Logger, snapshot Meta) (string, error) {
+	snapshotHash := md5.New()
+	err := hashFile(log, snapshotHash, snapshot.CurrentStateSnapshotFile)
+	if err != nil {
+		return "", err
+	}
+
+	err = hashFile(log, snapshotHash, snapshot.HistorySnapshotFile)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(snapshotHash.Sum(nil)), nil
+}
+
+func hashFile(log *logging.Logger, hash hash.Hash, filepath string) error {
+	filePath := filepath
+	file, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := file.Close()
+		if err != nil {
+			log.Errorf("failed to close file %s after hashing:%w", filePath, err)
+		}
+	}()
+	_, err = io.Copy(hash, file)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func WriteLockFileName(chainID string, height int64) string {
+	return fmt.Sprintf("%s-%d.writelock", chainID, height)
+}
+
+func (b *Service) createUncompressedDataDirectory(uncompressedDataDir string) (string, error) {
+	dir := filepath.Join(b.snapshotsPath, uncompressedDataDir)
```

Comment:
For all things path related, would be good to have those things covered with the paths libs I suppose. cc @ValentinTrinque